### PR TITLE
Tweaking rejit heuristics

### PR DESCRIPTION
**Pull request by rajatd**
Our current threshold of call-count-to-bailout-count ratio beyond which a rejit is triggered, is 20. Also, we don’t have such a ratio for loop bodies, we just rejit after the second bailout for the same bailout record. I had been looking into tuning the ratio for function rejits and also to make loop body rejits heuristics based instead of a hard bailout limit of 2.

Making rejits less aggressive should generally help with CPU time for websites. However, making them too conservative starts to show regressions on benchmarks due to increased number of bailouts. From my experiments, sweet spot for call-count to bailout-count ration for rejit seems to 10.

Loop body rejits:
1. Implemented a counter for loop bodies to keep track of how many iterations we run in the jitted code before bailing out of a jitted loop body. Then, I use the ratio of jitted-iterations-count to bailout-count to trigger a rejit. Settling on a value of 50 as the threshold for this ratio.
2. Ported some preemptive rejit strategies that we use for functions, to the loop body rejitting code.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/1029)

### CODE REVIEW

**Comment by rajatd** [comment]
@dotnet-bot test this please

**Comment by rajatd** [comment]
@dotnet-bot test Windows x64_test please

**Comment by rajatd** [comment]
@LouisLaf @pleath @curtisman @abchatra , please take a look.
FYI, @ThomsonTan

**Comment by rajatd** [comment]
A bunch of CPU time improvements in websites- http://ieperf-web-01/perfresults/Comparison?testRunId=411792&baseRunId=411787&lab=IEPERF&getRelatedRunData=false&priorities=0&categories=failures,warnings,improvements,no-change&search=

File: lib/Runtime/Base/FunctionBody.h [hunk 1]
```diff
@@ -984,8 +983,10 @@ namespace Js
     {
     public:
         LoopHeader* loopHeader;
+        uint jittedLoopIterationsSinceLastBailout; // number of times the loop iterated in the jitted code before bailing out 
+        uint totalJittedLoopIterations; // total number of times the loop has iterated in the jitted code for this entry point for a particular invocation of the loop
         LoopEntryPointInfo(LoopHeader* loopHeader, Js::JavascriptLibrary* library, void* validationCookie) :
-            loopHeader(loopHeader), mIsTemplatizedJitMode(false),EntryPointInfo(nullptr, library, validationCookie, /*threadContext*/ nullptr, /*isLoopBody*/ true)
+            loopHeader(loopHeader), jittedLoopIterationsSinceLastBailout(0), totalJittedLoopIterations(0), mIsTemplatizedJitMode(false),EntryPointInfo(nullptr, library, validationCookie, /*threadContext*/ nullptr, /*isLoopBody*/ true)
```
Comment (ThomsonTan) [inline lib/Runtime/Base/FunctionBody.h]: Line too long, put each field initialization into a separate line?

File: lib/Runtime/Base/FunctionBody.cpp [hunk 2]
```diff
@@ -3243,8 +3243,16 @@ namespace Js
         }
 #endif
         Assert(((LoopEntryPointInfo*) entryPointInfo)->loopHeader == loopHeader);
-        Assert(reinterpret_cast<void*>(entryPointInfo->jsMethod) == nullptr);
-        entryPointInfo->jsMethod = entryPoint;
+        Assert(entryPointInfo->jsMethod == nullptr);
+        entryPointInfo->jsMethod = (void*)entryPoint;
```
Comment (ThomsonTan) [inline lib/Runtime/Base/FunctionBody.cpp]: Use reinterpret_cast as convention in C++?

File: lib/Runtime/Base/FunctionBody.h (hunk 1 above)
Comment (rajatd) [inline lib/Runtime/Base/FunctionBody.h]: Sure.

File: lib/Runtime/Base/FunctionBody.cpp (hunk 2 above)
Comment (rajatd) [inline lib/Runtime/Base/FunctionBody.cpp]: Ah.. conflict resolve issue. Don't know why it didn't show up in my diff. Thanks for catching.

---
In reply to: [65243518](https://github.com/Microsoft/ChakraCore/pull/1029#discussion_r65243518) [](ancestors = 65243518)

File: lib/Backend/BailOut.cpp [hunk 3]
```diff
@@ -2263,15 +2263,62 @@ void BailOutRecord::ScheduleLoopBodyCodeGen(Js::ScriptFunction * function, Js::S
...
+            // Note you need to include only bailout kinds which don't disable the entire optimizations.
+            REJIT_KIND_TESTTRACE(bailOutKind, _u("Bailout from loop: function: %s, loopNumber: %d, jittedLoopIterations: %d, bailOutKindName: (%S), reJitReason: %S\r\n"),
+                function->GetFunctionBody()->GetDisplayName(), executeFunction->GetLoopNumber(loopHeader), totalJittedLoopIterations,
+                ::GetBailOutKindName(bailOutKind), RejitReasonNames[rejitReason]);
+
+            bailOutRecordNotConst->bailOutCount = 0;
+            totalJittedLoopIterations = 0;
+            break;
+        }
+        default: break;
```
Comment (ThomsonTan) [inline lib/Backend/BailOut.cpp]: When could hit this default?

Comment (rajatd) [inline lib/Backend/BailOut.cpp]: 'default' will hit for every other bail out. I think the strategy here has been to add more bailout kinds to this switch as we discover a problem of never (or very late) rejitting with them.

**Comment by ThomsonTan** [comment]
:shipit:

**Pull request merged**